File: backend/app/schemas/visualization.py
"""Pydantic response schemas for the visualization API endpoint.

Defines the complete visualization payload structure: facts with source spans,
claims with elements, mappings, gaps, and messages. Used by the
GET /api/v1/analysis/{intake_id}/visualization endpoint.
"""

from __future__ import annotations

from pydantic import BaseModel


class VisualizationSourceSpan(BaseModel):
    """A character/time/page range linking a fact back to its source message."""

    message_id: int
    start_char: int
    end_char: int
    page_number: int | None = None
    paragraph_index: int | None = None
    timestamp_start_sec: float | None = None
    timestamp_end_sec: float | None = None


class VisualizationFact(BaseModel):
    """An extracted fact with its source provenance spans."""

    id: int
    assertion_text: str
    fact_type: str
    confidence: float
    source_spans: list[VisualizationSourceSpan] = []


class VisualizationElement(BaseModel):
    """A required element of a legal claim."""

    id: int
    element_name: str
    element_description: str | None = None
    is_satisfied: bool
    satisfaction_confidence: float | None = None


class VisualizationClaim(BaseModel):
    """A legal claim with its constituent elements."""

    id: int
    claim_name: str
    claim_type: str
    jurisdiction: str | None = None
    confidence: float
    rationale: str | None = None
    elements: list[VisualizationElement] = []


class VisualizationMapping(BaseModel):
    """A fact-to-claim-element mapping with confidence score."""

    id: int
    fact_id: int
    claim_id: int
    element_id: int | None = None
    confidence: float
    mapping_rationale: str | None = None


class VisualizationGap(BaseModel):
    """An identified gap in the analysis."""

    id: int
    gap_type: str
    claim_id: int | None = None
    element_id: int | None = None
    description: str
    priority: int
    status: str


class VisualizationMessage(BaseModel):
    """A source message with decoded content for source span rendering."""

    id: int
    content: str
    sender_type: str


class VisualizationResponse(BaseModel):
    """Complete visualization payload for a single analysis run."""

    run_id: int
    status: str
    facts: list[VisualizationFact] = []
    claims: list[VisualizationClaim] = []
    mappings: list[VisualizationMapping] = []
    gaps: list[VisualizationGap] = []
    messages: list[VisualizationMessage] = []

File: frontend/src/test/fixtures/visualization.ts
/**
 * Mock visualization data for tests and MSW handlers.
 *
 * Realistic landlord-tenant dispute scenario:
 * - 3 facts about habitability issues
 * - 2 claims (breach of warranty of habitability, wrongful eviction)
 * - 4 elements across both claims
 * - 5 mappings linking facts to elements
 * - 2 gaps (missing evidence)
 * - 3 messages (consumer narrative, professional question, consumer response)
 */

import type { VisualizationData } from '@/features/visualization/types'

export const mockVisualizationData: VisualizationData = {
  run_id: 1,
  status: 'completed',

  facts: [
    {
      id: 101,
      assertion_text: 'Landlord has refused to fix the broken heater since November 2025',
      fact_type: 'assertion',
      confidence: 0.92,
      source_spans: [
        {
          message_id: 301,
          start_char: 0,
          end_char: 64,
          page_number: null,
          paragraph_index: null,
          timestamp_start_sec: null,
          timestamp_end_sec: null,
        },
      ],
    },
    {
      id: 102,
      assertion_text: 'Apartment has visible mold in the bathroom and kitchen',
      fact_type: 'condition',
      confidence: 0.88,
      source_spans: [
        {
          message_id: 303,
          start_char: 20,
          end_char: 74,
          page_number: null,
          paragraph_index: null,
          timestamp_start_sec: null,
          timestamp_end_sec: null,
        },
      ],
    },
    {
      id: 103,
      assertion_text: 'Landlord served a 30-day notice to vacate after tenant complained',
      fact_type: 'event',
      confidence: 0.85,
      source_spans: [
        {
          message_id: 301,
          start_char: 120,
          end_char: 184,
          page_number: null,
          paragraph_index: null,
          timestamp_start_sec: null,
          timestamp_end_sec: null,
        },
      ],
    },
  ],

  claims: [
    {
      id: 201,
      claim_name: 'Breach of Warranty of Habitability',
      claim_type: 'identified',
      jurisdiction: 'California',
      confidence: 0.85,
      rationale:
        'Landlord failed to maintain habitable conditions by not repairing heating and allowing mold growth',
      elements: [
        {
          id: 401,
          element_name: 'Defective Condition',
          element_description:
            'The rental unit has a condition that substantially impairs health and safety',
          is_satisfied: true,
          satisfaction_confidence: 0.90,
        },
        {
          id: 402,
          element_name: 'Notice to Landlord',
          element_description:
            'Tenant gave landlord reasonable notice of the defective condition',
          is_satisfied: false,
          satisfaction_confidence: 0.30,
        },
        {
          id: 403,
          element_name: 'Reasonable Time to Repair',
          element_description: 'Landlord was given a reasonable time to make repairs',
          is_satisfied: true,
          satisfaction_confidence: 0.75,
        },
      ],
    },
    {
      id: 202,
      claim_name: 'Wrongful Eviction / Retaliatory Eviction',
      claim_type: 'potential',
      jurisdiction: 'California',
      confidence: 0.60,
      rationale:
        'Eviction notice served shortly after tenant complaints suggests retaliation',
      elements: [
        {
          id: 404,
          element_name: 'Protected Activity',
          element_description: 'Tenant engaged in a legally protected activity (complaint)',
          is_satisfied: true,
          satisfaction_confidence: 0.82,
        },
      ],
    },
  ],

  mappings: [
    {
      id: 501,
      fact_id: 101,
      claim_id: 201,
      element_id: 401,
      confidence: 0.88,
      mapping_rationale: 'Broken heater constitutes defective condition affecting habitability',
    },
    {
      id: 502,
      fact_id: 102,
      claim_id: 201,
      element_id: 401,
      confidence: 0.82,
      mapping_rationale: 'Mold growth in multiple rooms is a health hazard and defective condition',
    },
    {
      id: 503,
      fact_id: 101,
      claim_id: 201,
      element_id: 403,
      confidence: 0.75,
      mapping_rationale: 'Heater broken since November 2025 indicates landlord had reasonable time',
    },
    {
      id: 504,
      fact_id: 103,
      claim_id: 202,
      element_id: 404,
      confidence: 0.78,
      mapping_rationale: 'Eviction notice after complaint suggests retaliatory motive',
    },
    {
      id: 505,
      fact_id: 101,
      claim_id: 202,
      element_id: 404,
      confidence: 0.65,
      mapping_rationale: 'Complaint about heater is protected activity that preceded eviction',
    },
  ],

  gaps: [
    {
      id: 601,
      gap_type: 'unsupported_element',
      claim_id: 201,
      element_id: 402,
      description: 'No evidence that tenant formally notified landlord of defective conditions',
      priority: 1,
      status: 'open',
    },
    {
      id: 602,
      gap_type: 'weak_mapping',
      claim_id: 202,
      element_id: null,
      description:
        'Timeline between complaint and eviction notice needs clarification for retaliation claim',
      priority: 2,
      status: 'open',
    },
  ],

  messages: [
    {
      id: 301,
      content:
        'Landlord has refused to fix the broken heater since November 2025. I have been asking repeatedly but nothing has been done. Then last week he served me a 30-day notice to vacate after I complained about the conditions.',
      sender_type: 'consumer',
    },
    {
      id: 302,
      content:
        'Can you tell me more about the conditions in the apartment? Are there other maintenance issues beyond the heater?',
      sender_type: 'professional',
    },
    {
      id: 303,
      content:
        'Yes, there is also visible mold in the bathroom and kitchen that has been growing for months. The landlord knows about it but has not done anything.',
      sender_type: 'consumer',
    },
  ],
}
